detectors/azure/vm/vm.go

```diff
@@ -0,0 +1,137 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package vm // import "go.opentelemetry.io/contrib/detectors/azure/vm"
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"net/http"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/resource"
+	semconv "go.opentelemetry.io/otel/semconv/v1.24.0"
+)
+
+const (
+	defaultAzureVMMetadataEndpoint = "http://169.254.169.254/metadata/instance/compute?api-version=2021-12-13&format=json"
+)
+
+type config struct {
+	endpoint string
+}
+
+func newConfig(options ...Option) config {
+	c := config{defaultAzureVMMetadataEndpoint}
+	for _, option := range options {
+		c = option.apply(c)
+	}
+
+	return c
+}
+
+// Option applies an Azure VM detector configuration option.
+type Option interface {
+	apply(config) config
+}
+
+type optionFunc func(config) config
+
+func (fn optionFunc) apply(c config) config {
+	return fn(c)
+}
+
+// WithEndpoint sets the endpoint for obtaining a Azure instance metadata JSON.
+func WithEndpoint(e string) Option {
+	return optionFunc(func(c config) config {
+		c.endpoint = e
+
+		return c
+	})
+}
+
+type resourceDetector struct {
+	endpoint string
+}
+
+type vmMetadata struct {
+	VMId       *string `json:"vmId"`
+	Location   *string `json:"location"`
+	ResourceId *string `json:"resourceId"`
+	Name       *string `json:"name"`
+	VMSize     *string `json:"vmSize"`
+	OsType     *string `json:"osType"`
+	Version    *string `json:"version"`
+}
+
+// New returns a [resource.Detector] that will detect Azure VM resources.
+func New(opts ...Option) resource.Detector {
+	c := newConfig(opts...)
+	return &resourceDetector{c.endpoint}
+}
+
+// Detect detects associated resources when running on an Azure VM.
+func (detector *resourceDetector) Detect(ctx context.Context) (*resource.Resource, error) {
```

Comment:
Detect should not return an error if it is not running on an azure VM.  It should only return errors if it is running on an azure VM, but fails for some reason to detect or process the response.  https://github.com/open-telemetry/opentelemetry-specification/blob/f3aeadbd152ce50c4d2cc9bc94902613bb49cffd/specification/resource/sdk.md?plain=1#L115

Reply:
Done. If the metadata server can't be reached then an empty resource is returned. If it returns invalid JSON, then an error is returned.